File: src/main/java/com/pancisin/bookster/rest/controllers/EventBotController.java
package com.pancisin.bookster.rest.controllers;

import java.security.Principal;
import java.util.UUID;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.messaging.handler.annotation.DestinationVariable;
import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.pancisin.bookster.components.EventBotService;
import com.pancisin.bookster.models.EventBot;
import com.pancisin.bookster.models.EventBotRun;
import com.pancisin.bookster.models.enums.BotRunState;
import com.pancisin.bookster.repository.EventBotRepository;

@RestController
//@PreAuthorize("hasRole('SUPERADMIN')")
@RequestMapping("/api/event-bot/{bot_id}")
public class EventBotController {

	@Autowired
	private EventBotRepository eventBotRepository;

	@Autowired
	private EventBotService eventBotService;
	
	@Autowired
	private SimpMessagingTemplate webSocket;
	
	@GetMapping
	public ResponseEntity<?> getEventBot(@PathVariable UUID bot_id) {
		return ResponseEntity.ok(eventBotRepository.findOne(bot_id));
	}

	@DeleteMapping
	public ResponseEntity<?> deleteMapping(@PathVariable UUID bot_id) {
		eventBotRepository.delete(bot_id);
		return ResponseEntity.ok("success");
	}

	@PatchMapping("/toggle-active")
	public ResponseEntity<?> toggleActive(@PathVariable UUID bot_id) {
		EventBot stored = eventBotRepository.findOne(bot_id);
		stored.setActive(!stored.isActive());
		return ResponseEntity.ok(eventBotRepository.save(stored));
	}
	
	@MessageMapping("/bot/{bot_id}/run")
	public void runEventBot(@DestinationVariable("bot_id") UUID bot_id, Principal principal) {
		EventBot stored = eventBotRepository.findOne(bot_id);

		webSocket.convertAndSendToUser(principal.getName(), "/queue/page.bots", new EventBotRun(stored, BotRunState.RUNNING));
		
		EventBotRun run = eventBotService.run(stored);
		webSocket.convertAndSendToUser(principal.getName(), "/queue/page.bots", run);
	}
}

File: src/main/kotlin/com/pancisin/bookster/rest/controllers/v1/EventBotController.kt
package com.pancisin.bookster.rest.controllers.v1

import java.security.Principal
import java.util.UUID

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.messaging.handler.annotation.DestinationVariable
import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.messaging.simp.SimpMessagingTemplate

import com.pancisin.bookster.services.EventBotService
import com.pancisin.bookster.model.EventBot
import com.pancisin.bookster.model.BotRun
import com.pancisin.bookster.model.enums.BotRunState
import com.pancisin.bookster.repository.EventBotRepository
import com.pancisin.bookster.repository.BotRunRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.web.bind.annotation.*
import javax.validation.Valid

@RestController
// @PreAuthorize("hasRole('SUPERADMIN')")
@RequestMapping("/api/v1/event-bot/{bot_id}")
class EventBotController {

  @Autowired
  lateinit var eventBotRepository: EventBotRepository

  @Autowired
  lateinit var botRunRepository: BotRunRepository

  @Autowired
  lateinit var eventBotService: EventBotService

  @Autowired
  lateinit var webSocket: SimpMessagingTemplate

  @GetMapping
  fun getEventBot(@PathVariable bot_id: UUID) = ResponseEntity.ok(eventBotRepository.findOne(bot_id))

  @DeleteMapping
  fun deleteMapping(@PathVariable bot_id: UUID): ResponseEntity<String> {
    eventBotRepository.delete(bot_id)
    return ResponseEntity.ok("success")
  }

  @PutMapping
  fun putEventBot(@PathVariable bot_id: UUID, @RequestBody @Valid eventBot: EventBot): ResponseEntity<EventBot>? {
    val stored = eventBotRepository.findOne(bot_id)
    return ResponseEntity.ok(eventBotRepository.save(stored.apply {
      name = eventBot.name
      fbPageId = eventBot.fbPageId
    }))
  }

  @PatchMapping("/toggle-active")
  fun toggleActive(@PathVariable bot_id: UUID): ResponseEntity<EventBot> {
    val stored = eventBotRepository.findOne(bot_id).apply {
      active = !active
    }
    return ResponseEntity.ok(eventBotRepository.save(stored))
  }

  @GetMapping("/run/{page}/{size}")
  fun getRuns(
    @PathVariable bot_id: UUID,
    @PathVariable page: Int,
    @PathVariable size: Int
  ): ResponseEntity<Page<BotRun>> {
    return ResponseEntity.ok(botRunRepository.getByEventBot(bot_id, PageRequest(page, size, Sort.Direction.DESC, "date")))
  }

  @MessageMapping("/bot/{bot_id}/run")
  fun runEventBot(@DestinationVariable("bot_id") bot_id: UUID, principal: Principal) {
    val stored = eventBotRepository.findOne(bot_id)

    webSocket.convertAndSendToUser(principal.name, "/queue/page.bots", BotRun(stored, BotRunState.RUNNING))

    val run = eventBotService.run(stored)
    webSocket.convertAndSendToUser(principal.name, "/queue/page.bots", run)
  }
}
